src/ADAL.PCL/AcquireTokenHandlerBase.cs

```diff
@@ -105,46 +109,55 @@ internal abstract class AcquireTokenHandlerBase
         protected UserIdentifierType UserIdentifierType { get; set; }
 
         protected bool LoadFromCache { get; set; }
-        
+
         protected bool StoreToCache { get; set; }
 
+
         public async Task<AuthenticationResult> RunAsync()
         {
             bool notifiedBeforeAccessCache = false;
-
+            AuthenticationResultEx extendedLifetimeResultEx = null;
             try
             {
                 await this.PreRunAsync();
-
                 
+
                 if (this.LoadFromCache)
                 {
+                    
                     this.NotifyBeforeAccessCache();
                     notifiedBeforeAccessCache = true;
 
                     ResultEx = this.tokenCache.LoadFromCache(CacheQueryData, this.CallState);
                     this.ValidateResult();
+                    extendedLifetimeResultEx = ResultEx;
 
-                    if (ResultEx != null && ResultEx.Result.AccessToken == null && ResultEx.RefreshToken != null)
+                    if (ResultEx != null && ResultEx.Result!=null)
                     {
-                        ResultEx = await this.RefreshAccessTokenAsync(ResultEx);
-                        if (ResultEx != null && ResultEx.Exception == null)
+                        if ((ResultEx.Result.AccessToken == null && ResultEx.RefreshToken != null) ||
+                            (ResultEx.Result.ExtendedLifeTimeToken && ResultEx.RefreshToken!=null))
```

Comment:
can you add comments over the updated if sections? comment should address why and not the what of the flow.